src/script/E2EIdentity/E2EIdentityEnrollment.ts

```diff
@@ -252,66 +241,67 @@ export class E2EIHandler extends TypedEventEmitter<Events> {
+      const {challenge, keyAuth} = challengeData;
+      OIDCServiceStore.store.targetURL(challenge.target);
+      userData = await this.oidcService.authenticate(keyAuth, challenge.url, silent);
+    } else {
+      // If there is no challengeData, that means we have already authenticated the user and we just need to get the userdata
+      userData = await this.oidcService.getUser();
+    }
+    if (!userData) {
+      throw new Error('No user data received');
+    }
+    return userData.id_token;
+  }
+
+  private async enroll(attemptSilentAuth = false) {
     if (!this.config) {
       throw new Error('Trying to enroll for E2EI without initializing the E2EIHandler');
     }
     try {
       // Notify user about E2EI enrolment in progress
-      this.currentStep = E2EIHandlerStep.ENROLL;
       const isCertificateRenewal = await hasActiveCertificate();
       this.showLoadingMessage();
 
-      if (!userData) {
-        // If the enrolment is in progress, we need to get the id token from the oidc service, since oauth should have already been completed
-        if (await this.coreE2EIService.isEnrollmentInProgress()) {
-          // The redirect-url which is needed inside the OIDCService is stored in the OIDCServiceStore previously
-          this.oidcService = this.createOIDCService();
-          userData = await this.oidcService.handleAuthentication();
-          if (!userData) {
-            throw new Error('Received no user data from OIDC service');
-          }
-        }
-      }
-      const oAuthIdToken = userData?.id_token;
-
       const displayName = this.userState.self()?.name();
       const handle = this.userState.self()?.username();
       const teamId = this.userState.self()?.teamId;
       // If the user has no username or handle, we cannot enroll
       if (!displayName || !handle || !teamId) {
         throw new Error('Username, handle or teamId not found');
       }
-      const enrollmentState = await this.core.enrollE2EI({
+
+      await this.core.enrollE2EI({
         discoveryUrl: this.config.discoveryUrl,
         displayName,
         handle,
         teamId,
-        oAuthIdToken,
+        getOAuthToken: async authenticationChallenge => {
+          return this.getOAuthToken(attemptSilentAuth, authenticationChallenge);
+        },
         certificateTtl: this.certificateTtl,
       });
-      // If the data is false or we dont get the ACMEChallenge, enrolment failed
-
-      if (enrollmentState.status === 'authentication') {
-        // If the data is authentication flow data, we need to kick off the oauth flow to get an oauth token
-        const {challenge, keyAuth} = enrollmentState.authenticationChallenge;
-        OIDCServiceStore.store.targetURL(challenge.target);
-        this.oidcService = this.createOIDCService();
-        await this.oidcService.authenticate(keyAuth, challenge.url);
-      }
```

Comment:
This is also included in the `getOAuthToken` method (the webapp doesn't really know which step we are at (before or after a redirect)